Instructor: search responses: handle legacy search documents

v6.4.0

The following search does not work. InstructorSearchPageUiTest fails on the live server as well. This was working in v6.3.0.

![image](https://user-images.githubusercontent.com/1673303/37551669-6cbb6606-29df-11e8-9b5e-fcb8832e8a16.png)

>teammates.common.util.Logger info: Query: courseId:(CS2103-Aug2011 OR CS2103-Aug2011-Java OR CS2103-Aug2011-SC OR CS2103-Aug2012 OR CS2103-Aug2013 OR CS2103-Jan2014 OR CS2103T-Aug2011 OR CS2103T-Aug2012 OR CS3216Workshop-Aug2013 OR CS4217-Jan2012 OR CS4217-Jan2013 OR ENG1201-2013-S1 OR GEM1909-Aug2013-Mon OR GEM1909-Aug2013-Thu OR GEM1909-Jan2013 OR Newcourseafterinstructor OR TE3201-Jan2012 OR TE3201-Jan2014 OR big-team-course OR demo-create-course-2 OR test-activation OR testcourse-3 OR testcourse2 OR testcourse3 OR CS2103-Jan2017 OR CS2103-Aug2014 OR CS2103-Aug2015 OR CS2103-Aug2016 OR CS2103-Aug2017-PhaseA OR CS2103-Aug2017-PhaseB OR CS2103-Jan2015 OR CS2103-Jan2016 OR CS2103-Jan2018-PhaseA OR CS2103-Jan2018-PhaseB OR CS3216-Aug-2014 OR CS3281-2016 OR CS3281-2017 OR CS3281-2018 OR TE3201-Jan2015 OR TE3201-Jan2016 OR TE3201-Jan2018 OR damith.gma-demo1 OR damith.gma-demo2 OR damith.gma-demo3 OR emptycourse OR sbrspj01 OR temp-course OR temp-for-testing OR test-archiving) AND searchableText:("hard") (Logger.java:34)
2018-03-17 04:32:58.292 UTC
teammates.common.util.Logger severe: Unexpected exception caught by ControllerServlet :  (Logger.java:52)
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 21 column 17 path $.createdAt
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:853)
	at com.google.gson.Gson.fromJson(Gson.java:802)
	at teammates.common.util.JsonUtils.fromJson(JsonUtils.java:78)
	at teammates.storage.search.FeedbackResponseCommentSearchDocument.fromResults(FeedbackResponseCommentSearchDocument.java:259)
	at teammates.storage.api.FeedbackResponseCommentsDb.search(FeedbackResponseCommentsDb.java:314)
	at teammates.logic.core.FeedbackResponseCommentsLogic.searchFeedbackResponseComments(FeedbackResponseCommentsLogic.java:160)
	at teammates.logic.api.Logic.searchFeedbackResponseComments(Logic.java:1986)
	at teammates.ui.controller.InstructorSearchPageAction.execute(InstructorSearchPageAction.java:51)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:476)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:76)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:51)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1772)
	at com.googlecode.objectify.ObjectifyFilter.doFilter(ObjectifyFilter.java:48)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.runtime.jetty9.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:48)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:582)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at com.google.apphosting.runtime.jetty9.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:297)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:534)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
	at com.google.apphosting.runtime.jetty9.RpcConnection.handle(RpcConnection.java:202)
	at com.google.apphosting.runtime.jetty9.RpcConnector.serviceRequest(RpcConnector.java:81)
	at com.google.apphosting.runtime.jetty9.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:108)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:686)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:648)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:618)
	at com.google.apphosting.runtime.JavaRuntime$NullSandboxRequestRunnable.run(JavaRuntime.java:812)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:274)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 21 column 17 path $.createdAt
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:385)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:213)
	... 50 more

--- a/src/client/java/teammates/client/scripts/DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat.java
+++ b/src/client/java/teammates/client/scripts/DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat.java
@@ -1,10 +1,8 @@
 package teammates.client.scripts;
 
 import java.io.IOException;
-import java.text.DateFormat;
-import java.text.ParseException;
-import java.text.SimpleDateFormat;
-import java.util.Locale;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
 
 import com.google.appengine.api.search.Document;
 import com.google.gson.JsonParser;
@@ -13,20 +11,21 @@ import teammates.common.datatransfer.attributes.FeedbackResponseCommentAttribute
 import teammates.common.util.Const;
 
 /**
- * Script to fix the date format in old {@link FeedbackResponseCommentSearchDocument}s created before V5.93.
+ * Script to fix the date format in old {@link teammates.storage.search.FeedbackResponseCommentSearchDocument}s
+ * created before V6.4.0.
  *
- * <p>Before V5.93, we used GSON's default JSON serializer, which output the dates in en-US format.
- * From V5.93 onwards, we have been using our own JSON serializer which writes dates in our prescribed format.
+ * <p>Before V6.4.0, we used {@link java.util.Date} fields for timestamps. These are serialized using the format
+ * {@link Const.SystemParams#DEFAULT_DATE_TIME_FORMAT}.
+ * From V6.4.0 onwards, these have been migrated to {@link Instant} fields. These are serialized using
+ * {@link java.time.format.DateTimeFormatter#ISO_INSTANT}.
  * Comment search documents now exist in both formats, which the current codebase has to handle.</p>
  *
- * <p>This script migrates all comment search documents to the newer prescribed date format.</p>
+ * <p>This script migrates all comment search documents to the newer instant format.
+ * The @{@link teammates.common.datatransfer.attributes.FeedbackSessionAttributes#timeZone} field is also
+ * automatically migrated.</p>
  */
 public class DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat
         extends DataMigrationForFeedbackResponseCommentSearchDocument {
-
-    private static final DateFormat oldDateFormat =
-            DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.DEFAULT, Locale.US);
-    private static final DateFormat newDateFormat = new SimpleDateFormat(Const.SystemParams.DEFAULT_DATE_TIME_FORMAT);
 
     private static final JsonParser jsonParser = new JsonParser();
 
@@ -56,16 +55,11 @@ public class DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat
     @Override
     protected boolean isMigrationNeeded(FeedbackResponseCommentAttributes comment) {
         Document document = index.get(comment.getId().toString());
+        if (document == null) {
+            return false;
+        }
         String sampleDateString = extractSampleDateString(document);
-        if (isInDateFormat(sampleDateString, oldDateFormat)) {
-            return true;
-        }
-
-        if (!isInDateFormat(sampleDateString, newDateFormat)) {
-            println("Unrecognised date format (" + sampleDateString + ") for:\n" + comment);
-        }
-
-        return false;
+        return !isInInstantFormat(sampleDateString);
     }
 
     private String extractSampleDateString(Document document) {
@@ -73,11 +67,11 @@ public class DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat
         return jsonParser.parse(frcaJson).getAsJsonObject().getAsJsonPrimitive("createdAt").getAsString();
     }
 
-    private boolean isInDateFormat(String dateString, DateFormat dateFormat) {
+    private boolean isInInstantFormat(String dateString) {
         try {
-            dateFormat.parse(dateString);
+            Instant.parse(dateString);
             return true;
-        } catch (ParseException e) {
+        } catch (DateTimeParseException e) {
             return false;
         }
     }

--- a/src/main/java/teammates/common/util/JsonUtils.java
+++ b/src/main/java/teammates/common/util/JsonUtils.java
@@ -130,7 +130,7 @@ public final class JsonUtils {
             try {
                 return Instant.parse(element.getAsString());
             } catch (DateTimeParseException e) {
-                throw new JsonSyntaxException(element.getAsString(), e);
+                return new TeammatesDateAdapter().deserialize(element, type, context).toInstant();
             }
         }
     }
@@ -147,7 +147,7 @@ public final class JsonUtils {
             try {
                 return ZoneId.of(element.getAsString());
             } catch (DateTimeException e) {
-                throw new JsonSyntaxException(element.getAsString(), e);
+                return TimeHelper.convertToZoneId(element.getAsDouble());
             }
         }
     }

--- a/src/main/java/teammates/common/util/JsonUtils.java
+++ b/src/main/java/teammates/common/util/JsonUtils.java
@@ -75,6 +75,7 @@ public final class JsonUtils {
             return getTeammatesGson().fromJson(json, typeOfT);
         } catch (JsonSyntaxException e) {
             // some of the existing data does not use the prescribed date format
+            // TODO: remove once all search documents have been migrated
             return new Gson().fromJson(json, typeOfT);
         }
     }
@@ -113,6 +114,7 @@ public final class JsonUtils {
             try {
                 return dateFormat.parse(element.getAsString());
             } catch (ParseException e) {
+                // TODO: change to JsonParseException once all search documents have been migrated
                 throw new JsonSyntaxException(element.getAsString(), e);
             }
         }
@@ -130,6 +132,7 @@ public final class JsonUtils {
             try {
                 return Instant.parse(element.getAsString());
             } catch (DateTimeParseException e) {
+                // TODO: remove once all search documents have been migrated
                 return new TeammatesDateAdapter().deserialize(element, type, context).toInstant();
             }
         }
@@ -147,6 +150,7 @@ public final class JsonUtils {
             try {
                 return ZoneId.of(element.getAsString());
             } catch (DateTimeException e) {
+                // TODO: remove once all search documents have been migrated
                 return TimeHelper.convertToZoneId(element.getAsDouble());
             }
         }

